deepsubtree.go

```diff
@@ -0,0 +1,671 @@
+package iavl
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strings"
+
+	ics23 "github.com/confio/ics23/go"
+	dbm "github.com/tendermint/tm-db"
+)
+
+// Represents a IAVL Deep Subtree that can contain
+// a subset of nodes of an IAVL tree
+type DeepSubTree struct {
+	*MutableTree
+}
+
+type DSTNonExistenceProof struct {
+	*ics23.NonExistenceProof
+	LeftSiblingProof  *ics23.ExistenceProof
+	RightSiblingProof *ics23.ExistenceProof
+}
+
+// NewDeepSubTree returns a new deep subtree with the specified cache size, datastore, and version.
+func NewDeepSubTree(db dbm.DB, cacheSize int, skipFastStorageUpgrade bool, version int64) (*DeepSubTree, error) {
+	ndb := newNodeDB(db, cacheSize, nil)
+	head := &ImmutableTree{ndb: ndb, version: version}
+	mutableTree := &MutableTree{
+		ImmutableTree:            head,
+		lastSaved:                head.clone(),
+		orphans:                  map[string]int64{},
+		versions:                 map[int64]bool{},
+		allRootLoaded:            false,
+		unsavedFastNodeAdditions: make(map[string]*FastNode),
+		unsavedFastNodeRemovals:  make(map[string]interface{}),
+		ndb:                      ndb,
+		skipFastStorageUpgrade:   skipFastStorageUpgrade,
+	}
+	return &DeepSubTree{MutableTree: mutableTree}, nil
+}
+
+// Constructs a DSTNonExistenceProof using an ICS23 Non-Existence proof
+// and sibling node proofs. Returns the constructed DSTNonExistenceProof.
+func ConvertToDSTNonExistenceProof(
+	tree *MutableTree,
+	nonExistenceProof *ics23.NonExistenceProof,
+) (*DSTNonExistenceProof, error) {
+	dstNonExistenceProof := DSTNonExistenceProof{
+		NonExistenceProof: nonExistenceProof,
+	}
+	if nonExistenceProof.Left != nil {
+		leftSibling, err := tree.getSiblingNode(nonExistenceProof.Left.Key)
+		if err != nil {
+			return nil, err
+		}
+		dstNonExistenceProof.LeftSiblingProof, err = createExistenceProof(tree.ImmutableTree, leftSibling.key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if nonExistenceProof.Right != nil {
+		rightSibling, err := tree.getSiblingNode(nonExistenceProof.Right.Key)
+		if err != nil {
+			return nil, err
+		}
+		dstNonExistenceProof.RightSiblingProof, err = createExistenceProof(tree.ImmutableTree, rightSibling.key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &dstNonExistenceProof, nil
+}
+
+// Returns the sibling node of a leaf node with given key
+func (tree *ImmutableTree) getSiblingNode(key []byte) (*Node, error) {
```

Comment:
This looks kind of redundant. `recursiveGetSiblingNode` already returns node or error. I guess this may be a leftover after refactoring.

Reply:
Turned getSiblingNode into an exported method. Cleaner to have a wrapper method with fewer arguments and stay consistent with how Remove and Set have recursiveRemove and recursiveSet as well.